pallets/phala/src/compute/vault.rs

```diff
@@ -0,0 +1,511 @@
+		///
+		/// If the shutdown condition is met, all shares owned by the vault will be forced withdraw.
+		/// Note: This function doesn't guarantee no-op when there's error.
+		#[pallet::weight(0)]
+		#[frame_support::transactional]
+		pub fn check_and_maybe_force_withdraw(
+			origin: OriginFor<T>,
+			vault_pid: u64,
+		) -> DispatchResult {
+			ensure_signed(origin.clone())?;
+			let now = <T as registry::Config>::UnixTime::now()
+				.as_secs()
+				.saturated_into::<u64>();
+			let mut vault = ensure_vault::<T>(vault_pid)?;
+			base_pool::Pallet::<T>::try_process_withdraw_queue(&mut vault.basepool);
+			let grace_period = T::GracePeriod::get();
+			let mut releasing_stake = Zero::zero();
+			for pid in vault.invest_pools.iter() {
+				let stake_pool = ensure_stake_pool::<T>(*pid)?;
+				let withdraw_vec: VecDeque<_> = stake_pool
+					.basepool
+					.withdraw_queue
+					.iter()
+					.filter(|x| x.user == vault.basepool.pool_account_id)
+					.collect();
+				// the length of vec should be 1
+				for withdraw in withdraw_vec {
+					let nft_guard = base_pool::Pallet::<T>::get_nft_attr_guard(
+						stake_pool.basepool.cid,
+						withdraw.nft_id,
+					)?;
+					let price = stake_pool
+						.basepool
+						.share_price()
+						.ok_or(Error::<T>::VaultBankrupt)?;
+					releasing_stake += bmul(nft_guard.attr.shares, &price);
+				}
+			}
+			if vault.basepool.withdraw_queue.is_empty() {
+				VaultLocks::<T>::remove(vault_pid);
+			}
+			base_pool::pallet::Pools::<T>::insert(vault_pid, PoolProxy::Vault(vault.clone()));
+			if base_pool::Pallet::<T>::has_expired_withdrawal(
+				&vault.basepool,
+				now,
+				grace_period,
+				releasing_stake,
+			) {
+				for pid in vault.invest_pools.iter() {
+					let stake_pool = ensure_stake_pool::<T>(*pid)?;
+					let mut total_shares = Zero::zero();
+					let withdraw_vec: VecDeque<_> = stake_pool
+						.basepool
+						.withdraw_queue
+						.iter()
+						.filter(|x| x.user == vault.basepool.pool_account_id)
+						.collect();
+					// the length of vec should be 1
+					for withdraw in withdraw_vec {
+						let nft_guard = base_pool::Pallet::<T>::get_nft_attr_guard(
+							stake_pool.basepool.cid,
+							withdraw.nft_id,
+						)?;
+						total_shares += nft_guard.attr.shares;
+					}
+					pallet_uniques::Pallet::<T>::owned_in_collection(
+						&stake_pool.basepool.cid,
+						&vault.basepool.pool_account_id,
+					)
+					.for_each(|nftid| {
+						let property_guard = base_pool::Pallet::<T>::get_nft_attr_guard(
+							stake_pool.basepool.cid,
+							nftid,
+						)
+						.expect("get nft should not fail: qed.");
+						let property = &property_guard.attr;
+						total_shares += property.shares;
+					});
+					stake_pool_v2::Pallet::<T>::withdraw(
+						Origin::<T>::Signed(vault.basepool.owner.clone()).into(),
+						stake_pool.basepool.pid,
+						total_shares,
+						Some(vault_pid),
+					)?;
+				}
+				VaultLocks::<T>::insert(vault_pid, ());
```

Comment:
`VaultLocks` is only used in this function and never read. Do you missed to use it somewhere?

Reply:
it is used in stakepool, to forbid owner withdraw a small amount to cancel the automatic force shutdown